client.go

```diff
@@ -520,6 +524,107 @@ func (c *Client) ListImages(ctx context.Context, filters ...string) ([]Image, er
 	return images, nil
 }
 
+// Restore restores a container from a checkpoint
+func (c *Client) Restore(ctx context.Context, id, ref string, opts ...RestoreOpts) error {
+	checkpoint, err := c.GetImage(ctx, ref)
+	if err != nil {
+		if !errdefs.IsNotFound(err) {
+			return err
+		}
+		ck, err := c.Fetch(ctx, ref)
+		if err != nil {
+			return err
+		}
+		checkpoint = NewImage(c, ck)
+	}
+
+	store := c.ContentStore()
+	index, err := decodeIndex(ctx, store, checkpoint.Target())
+	if err != nil {
+		return err
+	}
+
+	// get image from annotation
+	imageName, ok := index.Annotations["image.name"]
+	if !ok {
+		return ErrCheckpointIndexImageNameNotFound
+	}
+
+	image, err := c.Pull(ctx, imageName, WithPullUnpack)
+	if err != nil {
+		return err
+	}
+
+	ctx, done, err := c.WithLease(ctx)
+	if err != nil {
+		return err
+	}
+	defer done(ctx)
+
+	// container options
+	copts := []NewContainerOpts{
+		WithNewSpec(oci.WithImageConfig(image)),
+		WithNewSnapshot(id, image),
+	}
+	topts := []NewTaskOpts{}
+	for _, o := range opts {
+		co, to, err := o(ctx, c, checkpoint, index)
+		if err != nil {
+			return err
+		}
+		copts = append(copts, co...)
+		topts = append(topts, to...)
+	}
+
+	ctr, err := c.NewContainer(ctx, id, copts...)
+	if err != nil {
+		return err
+	}
+
+	// apply rw layer
+	info, err := ctr.Info(ctx)
+	if err != nil {
+		return err
+	}
+
+	rw, err := GetIndexByMediaType(index, ocispec.MediaTypeImageLayerGzip)
+	if err != nil {
+		return err
+	}
+
+	mounts, err := c.SnapshotService(info.Snapshotter).Mounts(ctx, info.SnapshotKey)
+	if err != nil {
+		return err
+	}
+
+	if _, err := c.DiffService().Apply(ctx, *rw, mounts); err != nil {
+		return err
+	}
+
+	task, err := ctr.NewTask(ctx, cio.NewCreator(cio.WithStdio), topts...)
```

Comment:
Remove task handling from here, it's more complex with wait and signal handling and needs to be done by the caller.